core/src/services/swift/backend.rs

```diff
@@ -0,0 +1,335 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+
+use super::core::SwiftCore;
+use super::error::parse_error;
+use super::pager::SwiftPager;
+use super::writer::SwiftWriter;
+use crate::raw::*;
+use crate::*;
+
+const SWIFT_DEFAULT_ENDPOINT: &str = "http://127.0.0.1:8080";
+
+/// [OpenStack Swift](https://docs.openstack.org/api-ref/object-store/#)'s REST API support.
+#[doc = include_str!("docs.md")]
+#[derive(Default, Clone)]
+pub struct SwiftBuilder {
+    endpoint: Option<String>,
+    account_name: Option<String>,
+    container: Option<String>,
+    root: Option<String>,
+    token: Option<String>,
+}
+
+impl Debug for SwiftBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Builder");
+
+        ds.field("root", &self.root);
+        ds.field("endpoint", &self.endpoint);
+        ds.field("account_name", &self.account_name);
+        ds.field("container", &self.container);
+
+        if self.token.is_some() {
+            ds.field("token", &"<redacted>");
+        }
+
+        ds.finish()
+    }
+}
+
+impl SwiftBuilder {
+    /// Set the remote address of this backend
+    /// default to `http://127.0.0.1:8080`
+    ///
+    /// Endpoints should be full uri, e.g.
+    ///
+    /// - `https://openstack-controller.example.com:8080`
+    /// - `http://192.168.66.88:8080`
+    ///
+    /// If user inputs endpoint without scheme, we will
+    /// prepend `http://` to it.
```

Comment:
Prepend `https` for safer.

Reply:
Done, how about `WebHDFS`, it uses `http` for the same?

Reply:
Oh, that must be a bug. But we can't change that since it has been released.

Reply:
Ok, got it.